core/network/src/peer_manager/mod.rs

```diff
@@ -36,403 +48,482 @@ use log::{debug, error, info, warn};
-            self.remove_peer_addr(&pid, addr);
-        }
-    }
+    sessions: RwLock<HashSet<ArcSession>>,
+    peers:    RwLock<HashSet<ArcPeer>>,
+    chain:    RwLock<HashMap<Address, ArcPeer>>,
+
+    listen: RwLock<HashSet<PeerMultiaddr>>,
+}
 
-    pub fn remove_peer(&self, peer_id: &PeerId) {
-        self.connecting.write().remove(peer_id);
-        self.connected.write().remove(peer_id);
+impl Inner {
+    pub fn new() -> Self {
+        Inner {
+            whitelist: Default::default(),
 
-        let mut addr_pid = self.addr_pid.write();
-        if let Some(peer) = self.pool.write().remove(peer_id) {
-            self.user_pid.write().remove(peer.user_addr());
+            sessions: Default::default(),
+            peers:    Default::default(),
+            chain:    Default::default(),
 
-            for addr in peer.addrs().into_iter() {
-                addr_pid.remove(addr);
-            }
+            listen: Default::default(),
         }
     }
 
-    pub fn register_self(&self, peer_id: PeerId, pubkey: PublicKey) {
-        let user_addr = Peer::pubkey_to_addr(&pubkey);
-        let peer = Peer::new(peer_id.clone(), pubkey);
-
-        self.pool.write().insert(peer_id.clone(), peer);
-        self.user_pid.write().insert(user_addr, peer_id.clone());
-        self.connected.write().insert(peer_id);
+    pub fn add_listen(&self, multiaddr: PeerMultiaddr) {
+        self.listen.write().insert(multiaddr);
     }
 
-    pub fn peer_connected(&self, peer_id: &PeerId) -> bool {
-        self.connected.read().contains(peer_id)
+    pub fn listen(&self) -> HashSet<PeerMultiaddr> {
+        self.listen.read().clone()
     }
 
-    pub fn connect_peer(&self, peer_id: &PeerId) {
-        if self.connected.read().contains(peer_id) {
-            warn!("network: peer {:?} already connected", peer_id);
-            return;
-        }
-
-        self.connecting.write().insert(peer_id.clone());
+    pub fn remove_listen(&self, multiaddr: &PeerMultiaddr) {
+        self.listen.write().remove(multiaddr);
     }
 
-    pub fn set_connected(&self, peer_id: &PeerId, addr: Option<ConnectedAddr>) {
-        // Clean outbound connection
-        self.connecting.write().remove(peer_id);
-        self.connected.write().insert(peer_id.clone());
+    pub fn connected(&self) -> usize {
+        self.sessions.read().len()
+    }
 
-        let mut pool = self.pool.write();
-        if let Some(peer) = pool.get_mut(peer_id) {
-            peer.update_connect();
-            peer.set_connected_addr(addr);
+    pub fn add_peer(&self, peer: ArcPeer) {
+        self.peers.write().insert(peer.clone());
+        if let Some(chain_addr) = peer.owned_chain_addr() {
+            self.chain.write().insert(chain_addr, peer);
         }
     }
 
-    pub fn disconnect_peer(&self, peer_id: &PeerId) {
-        self.connecting.write().remove(peer_id);
-        self.connected.write().remove(peer_id);
+    pub(self) fn restore(&self, peers: Vec<ArcPeer>) {
```

Comment:
Rust has a lot of visibility, such as pub(crate), pub(self), and so on. I think this visibility adds an extra burden when review code, and most of the time pub and private are perfectly adequate.

What do you think?